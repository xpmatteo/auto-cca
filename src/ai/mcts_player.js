import { score } from "./score.js";

class TreeNode {
    constructor(game, parent = null) {
        this.game = game;
        this.parent = parent;
        this.score = 0;
        this.visits = 0;
        this.children = [];
    }

    size() {
        if (this.children.length === 0) {
            return 1;
        }
        return 1 + this.children.reduce((acc, child) => acc + child.size(), 0);
    }

    shape() {
        const result = [];
        function traverse(node, level) {
            if (!result[level]) {
                result[level] = 0;
            }
            result[level]++;
            for (const child of node.children) {
                traverse(child, level + 1);
            }
        }
        traverse(this, 0);
        return result;
    }

    toString() {
        const result = [];
        function traverse(node, level) {
            result.push(`${" ".repeat(level)}${node.score}/${node.visits}`);
            for (const child of node.children) {
                traverse(child, level + 1);
            }
        }
        traverse(this, 0);
        return result.join("\n");
    }
}

export class MctsPlayer {
    constructor(args) {
        this.args = args;
    }

    search() {
        const rootNode = new TreeNode(this.args.game);
        for (let i = 0; i < this.args.iterations; i++) {
            let node = this._select(rootNode);
            let score = this._simulate(node.game);
            this._backpropagate(node, score);
        }
        return rootNode;
    }

    _expand(node) {
        const game = node.game;
        const validCommands = game.validCommands();
        validCommands.forEach((command) => {
            const clone = game.clone();
            clone.executeCommand(command);
            const childNode = new TreeNode(clone, node);
            childNode.command = command;
            node.children.push(childNode);
        });
        return node.children[0];
    }

    _backpropagate(node, score) {
        while (node !== null) {
            node.score += score;
            node.visits++;
            node = node.parent;
        }
    }

    _select(node) {
        while (!node.game.isTerminal()) {
            if (node.children.length === 0) {
                return this._expand(node);
            } else {
                node = node.bestChild();
            }
        }
        return node;
    }

    _simulate(game) {
        return score(game, game.currentSide);
    }
}
