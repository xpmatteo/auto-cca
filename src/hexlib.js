"use strict";

// thanks https://www.redblobgames.com/grids/hexagons/#coordinates

class Hex {
    constructor(q, r, s=undefined) {
        if (s === undefined) {
            s = -q - r;
        }
        assertEquals(0, q + r + s, "q + r + s must be 0");
        this.q = q;
        this.r = r;
    }

    get s() { return - this.q - this.r }
};


class Orientation {
    // start_angle is in multiples of 60Â°
    constructor(f0_, f1_, f2_, f3_, b0_, b1_, b2_, b3_, start_angle_) {
        this.f0 = f0_; 
        this.f1 = f1_; 
        this.f2 = f2_; 
        this.f3 = f3_;
        this.b0 = b0_;
        this.b1 = b1_;
        this.b2 = b2_;
        this.b3 = b3_;
        this.start_angle = start_angle_;
    }
};

class Point {
    x; y;
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }
}

class Layout {    
    constructor(orientation, size, origin) {
        this.orientation = orientation;
        this.size = size;
        this.origin = origin;
    }
}

function hex_to_pixel(layout, hex) {
    let M = layout.orientation;
    let x = (M.f0 * hex.q + M.f1 * hex.r) * layout.size.x;
    let y = (M.f2 * hex.q + M.f3 * hex.r) * layout.size.y;
    return new Point(x + layout.origin.x, y + layout.origin.y);
}

function pixel_to_hex(layout, p) {
    const M = layout.orientation;
    let pt = new Point((p.x - layout.origin.x) / layout.size.x, 
                        (p.y - layout.origin.y) / layout.size.y);
    let q = M.b0 * pt.x + M.b1 * pt.y;
    let r = M.b2 * pt.x + M.b3 * pt.y;
    return hex_round(q, r, -q - r);
}

function hex_round(fracq, fracr, fracs) {
    let q = int(round(fracq));
    let r = int(round(fracr));
    let s = int(round(fracs));
    let q_diff = abs(q - fracq);
    let r_diff = abs(r - fracr);
    let s_diff = abs(s - fracs);
    if (q_diff > r_diff && q_diff > s_diff) {
        q = -r - s;
    } else if (r_diff > s_diff) {
        r = -q - s;
    } else {
        s = -q - r;
    }
    return new Hex(q, r, s);
}

const sqrt = Math.sqrt;

const layout_pointy
  = new Orientation(sqrt(3.0), sqrt(3.0) / 2.0, 0.0, 3.0 / 2.0,
                sqrt(3.0) / 3.0, -1.0 / 3.0, 0.0, 2.0 / 3.0,
                0.5);


function assertEquals(expected, actual, message = "Assertion failed") {
    if (expected != actual) {
        throw new Error(message + ": expected " + expected + ", got " + actual);
    }
}

function assertDeepEquals(expected, actual, message = "Assertion failed") {
    if (JSON.stringify(expected) != JSON.stringify(actual)) {
        throw new Error(message + ": expected " + expected + ", got " + actual);
    }
}

                
assertEquals(0, 1, "next: print pixel_to_hex for mouse click");

const test_layout = new Layout(layout_pointy, new Point(50, 60), new Point(10, 100));
assertDeepEquals(layout_pointy, test_layout.orientation);
assertDeepEquals(hex_to_pixel(test_layout, new Hex(0, 0, 0)), new Point(10, 100));
assertEquals(  0, new Hex(1,  -1, 0).s);
assertEquals(-30, new Hex(10, 20)   .s);
