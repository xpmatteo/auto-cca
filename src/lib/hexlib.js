"use strict";

const sqrt = Math.sqrt;
const abs = Math.abs;
const round = Math.round;
const trunc = Math.trunc;

// thanks https://www.redblobgames.com/grids/hexagons/implementation.html

class Hex {
    distances = new Map();

    constructor(q, r, s = undefined) {
        if (s === undefined) {
            s = -q - r;
        }
        assertEquals(0, q + r + s, "q + r + s must be 0");
        this.q = q;
        this.r = r;
    }

    get s() { return - this.q - this.r }

    toString() {
        return `[${this.q},${this.r}]`;
    }

    distance(other) {
        if (!this.distances.get(other)) {
            this.distances.set(other, hex_length(hex_subtract(this, other)));
        }
        return this.distances.get(other);
    }

    neighbors() {
        return HEX_DIRECTIONS.map(d => hex_add(this, d));
    }

    get northernNeighbors() {
        return [this.northEast, this.northWest];
    }

    get southernNeighbors() {
        return [this.southWest, this.southEast];
    }

    get southEast() {
        return hex_add(this, DIRECTION_SE);
    }

    get southWest() {
        return hex_add(this, DIRECTION_SW);
    }

    get northWest() {
        return hex_add(this, DIRECTION_NW);
    }

    get northEast() {
        return hex_add(this, DIRECTION_NE);
    }

    isNorthOf(otherHex) {
        return this.q - otherHex.q === 1 && this.r - otherHex.r === -2;
    };
    isNorthWestOf(otherHex) {
        return this.q - otherHex.q === -1 && this.r - otherHex.r === -1;
    }
};

// The only way to create a Hex is with hexOf(q, r)
const HEXES = [];
export function hexOf(q, r) {
    const key = q + r * 1000;
    if (HEXES[key] === undefined) {
        HEXES[key] = new Hex(q, r);
    }
    return HEXES[key];
}

const DIRECTION_WEST = hexOf(-1, 0);
const DIRECTION_NW =   hexOf(0, -1);
const DIRECTION_NE =   hexOf(1, -1);
const DIRECTION_SW =   hexOf(-1, 1);
const DIRECTION_SE =   hexOf(0, 1);
const DIRECTION_EAST = hexOf(1, 0);
const HEX_DIRECTIONS = [
    DIRECTION_EAST,
    DIRECTION_NW,
    DIRECTION_NE,
    DIRECTION_WEST,
    DIRECTION_SW,
    DIRECTION_SE
];

class Orientation {
    // start_angle is in multiples of 60Â°
    constructor(f0_, f1_, f2_, f3_, b0_, b1_, b2_, b3_, start_angle_) {
        this.f0 = f0_;
        this.f1 = f1_;
        this.f2 = f2_;
        this.f3 = f3_;
        this.b0 = b0_;
        this.b1 = b1_;
        this.b2 = b2_;
        this.b3 = b3_;
        this.start_angle = start_angle_;
    }
};

export class Point {
    x; y;
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }

    add(other) {
        return new Point(this.x + other.x, this.y + other.y);
    }
}

export class Layout {
    constructor(orientation, size, origin) {
        this.orientation = orientation;
        this.size = size;
        this.origin = origin;
    }
}

export function hex_to_pixel(layout, hex) {
    let M = layout.orientation;
    let x = (M.f0 * hex.q + M.f1 * hex.r) * layout.size.x;
    let y = (M.f2 * hex.q + M.f3 * hex.r) * layout.size.y;
    return new Point(x + layout.origin.x, y + layout.origin.y);
}

export function pixel_to_hex(layout, p) {
    const M = layout.orientation;
    let pt = new Point((p.x - layout.origin.x) / layout.size.x,
        (p.y - layout.origin.y) / layout.size.y);
    let q = M.b0 * pt.x + M.b1 * pt.y;
    let r = M.b2 * pt.x + M.b3 * pt.y;
    return hex_round(q, r, -q - r);
}

function hex_round(fracq, fracr, fracs) {
    let q = round(fracq);
    let r = round(fracr);
    let s = round(fracs);
    let q_diff = abs(q - fracq);
    let r_diff = abs(r - fracr);
    let s_diff = abs(s - fracs);
    if (q_diff > r_diff && q_diff > s_diff) {
        q = -r - s;
    } else if (r_diff > s_diff) {
        r = -q - s;
    } else {
        s = -q - r;
    }
    return hexOf(q, r, s);
}

function hex_add(a, b) {
    return hexOf(a.q + b.q, a.r + b.r, a.s + b.s);
}

function hex_subtract(a, b) {
    return hexOf(a.q - b.q, a.r - b.r, a.s - b.s);
}

function hex_length(hex) {
    return trunc((abs(hex.q) + abs(hex.r) + abs(hex.s)) / 2);
}

export const LAYOUT_POINTY = new Orientation(
    sqrt(3.0),
    sqrt(3.0) / 2.0,
    0.0,
    3.0 / 2.0,
    sqrt(3.0) / 3.0,
    -1.0 / 3.0,
    0.0,
    2.0 / 3.0,
    0.5);

function assertEquals(expected, actual, message = "Assertion failed") {
    if (expected !== actual) {
        throw new Error(message + ": expected " + expected + ", got " + actual);
    }
}

/**
 * Returns true if there is a line of sight from fromHex to toHex.
 * @param {Hex} toHex
 * @param {Hex} fromHex
 * @param {function(Hex):boolean} isBlocked
 * @returns {boolean}
 */
export function hasLineOfSight(toHex, fromHex, isBlocked = () => false) {
    if (fromHex.isNorthOf(toHex)) {
        // No unit in the game except heavy war machine is able to hit farther than 3 hexes.  And the
        // heavy war machine is not implemented yet ;-)
        return !isBlocked(hex_add(fromHex, DIRECTION_SW)) || !isBlocked(hex_add(fromHex, DIRECTION_SE));
    }
    if (toHex.isNorthOf(fromHex)) {
        return !isBlocked(hex_add(fromHex, DIRECTION_NW)) || !isBlocked(hex_add(fromHex, DIRECTION_NE));
    }
    if (fromHex.isNorthWestOf(toHex)) {
        return !isBlocked(hex_add(fromHex, DIRECTION_EAST)) || !isBlocked(hex_add(fromHex, DIRECTION_SE));
    }
    if (toHex.isNorthWestOf(fromHex)) {
        return !isBlocked(hex_add(fromHex, DIRECTION_NW)) || !isBlocked(hex_add(fromHex, DIRECTION_WEST));
    }
    const distance = fromHex.distance(toHex);
    const step = hex_subtract(toHex, fromHex);
    const stepSize = 1.0 / distance;
    for (let i = 1; i < distance; i++) {
        const stepHex = hex_round(fromHex.q + (step.q * stepSize * i),
            fromHex.r + (step.r * stepSize * i),
            fromHex.s + (step.s * stepSize * i));
        if (isBlocked(stepHex)) {
            return false;
        }
    }
    return true;
}
