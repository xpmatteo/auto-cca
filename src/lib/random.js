
export function randomShuffleArray(array) {
    for (let i = 0; i < array.length; i++) {
        const j = Math.floor(Math.random() * (array.length - i)) + i;
        [array[i], array[j]] = [array[j], array[i]];
    }
}

/**
 * @template T
 * @param {T[]} array
 * @returns {T}
 */
export function randomElement(array) {
    return array[Math.floor(Math.random() * array.length)];
}

const RANDOM_VALUES = [0.1, 0.3, 0.2, 0.7, 0.9, 0.6, 0.4, 0.8];
let nextRandom = 0;
export function fixedRandom() {
    return RANDOM_VALUES[nextRandom++ % 8];
}
