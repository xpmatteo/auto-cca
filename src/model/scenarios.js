import { hexOf } from "../lib/hexlib.js";
import GameStatus from "./game_status.js";
import { Side } from "./side.js";
import * as units from "./units.js";

export class Scenario {
    /**
     * @param side
     * @returns {Side}
     */
    opposingSide(side) {
        return side === this.sideNorth ? this.sideSouth : this.sideNorth;
    }

    /**
     * @param {Game} game
     * @returns {GameStatus}
     */
    gameStatus(game) {
        if (game.killedUnitsOfSide(Side.CARTHAGINIAN).length === this.pointsToWin) {
            return GameStatus.ROMAN_WIN;
        }
        if (game.killedUnitsOfSide(Side.ROMAN).length === this.pointsToWin) {
            return GameStatus.CARTHAGINIAN_WIN;
        }
        return GameStatus.ONGOING;
    }

    /**
     * @param {Board|Game} board
     */
    placeUnitsOn(board) {
        throw new Error("Subclass must implement");
    }
}

export class NullScenario extends Scenario {
    firstSide = Side.ROMAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.ROMAN;
    pointsToWin = 3;
    commandNorth = 1;
    commandSouth = 1;

    placeUnitsOn(board) {
    }

    gameStatus(board) {
        return GameStatus.ONGOING;
    }
}

export class MeleeScenario extends Scenario {
    firstSide = Side.ROMAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.ROMAN;
    pointsToWin = 3;
    commandNorth = 4;
    commandSouth = 3;

    placeUnitsOn(board) {
        // 1522
        board.placeUnit(hexOf(0, 5), new units.RomanHeavyInfantry());
        board.placeUnit(hexOf(1, 5), new units.RomanHeavyInfantry());
        board.placeUnit(hexOf(2, 5), new units.RomanHeavyInfantry());
        board.placeUnit(hexOf(3, 5), new units.RomanHeavyInfantry());

        // 6422
        // board.placeUnit(hexOf(1, 4), new units.RomanHeavyInfantry());
        // board.placeUnit(hexOf(2, 4), new units.RomanHeavyInfantry());
        // board.placeUnit(hexOf(3, 4), new units.RomanHeavyInfantry());
        // board.placeUnit(hexOf(4, 4), new units.RomanHeavyInfantry());

        board.placeUnit(hexOf(2, 3), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(4, 2), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(4, 3), new units.CarthaginianHeavyInfantry());
    }
}

export class OneToOneMeleeScenario extends Scenario {
    firstSide = Side.ROMAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.ROMAN;
    pointsToWin = 1;
    commandNorth = 3;
    commandSouth = 3;

    placeUnitsOn(board) {
        board.placeUnit(hexOf(2, 2), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(1, 4), new units.RomanLightBowsInfantry());
    }
}

export class TwoOnTwoMeleeScenario extends Scenario {
    firstSide = Side.ROMAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.ROMAN;
    pointsToWin = 2;
    commandNorth = 4;
    commandSouth = 3;

    placeUnitsOn(board) {
        board.placeUnit(hexOf(-1, 6), new units.RomanHeavyInfantry());
        board.placeUnit(hexOf(0, 6), new units.RomanHeavyInfantry());

        board.placeUnit(hexOf(0, 4), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(1, 4), new units.CarthaginianHeavyInfantry());
    }
}


export class AkragasScenario extends Scenario {
    firstSide = Side.SYRACUSAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.SYRACUSAN;
    pointsToWin = 5;
    commandNorth = 5;
    commandSouth = 6;

    placeUnitsOn(board) {
        board.placeUnit(hexOf(1, 1), new units.CarthaginianMediumInfantry());
        board.placeUnit(hexOf(2, 1), new units.CarthaginianAuxiliaInfantry());
        board.placeUnit(hexOf(3, 1), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(4, 1), new units.CarthaginianMediumInfantry());
        board.placeUnit(hexOf(6, 1), new units.CarthaginianAuxiliaInfantry());
        board.placeUnit(hexOf(7, 1), new units.CarthaginianAuxiliaInfantry());
        board.placeUnit(hexOf(8, 1), new units.CarthaginianAuxiliaInfantry());

        board.placeUnit(hexOf(1, 2), new units.CarthaginianHeavyChariot());
        board.placeUnit(hexOf(3, 2), new units.CarthaginianLightBowsInfantry());
        board.placeUnit(hexOf(7, 2), new units.CarthaginianLightInfantry());
        board.placeUnit(hexOf(9, 2), new units.CarthaginianHeavyChariot());
        board.placeUnit(hexOf(10, 1), new units.CarthaginianLightCavalry());

        function placeSyracusanUnit(hex, unit) {
            board.placeUnit(hex, unit);
            unit.side = Side.SYRACUSAN;
        }
        placeSyracusanUnit(hexOf(-2, 7), new units.RomanLightCavalry());
        placeSyracusanUnit(hexOf(-1, 7), new units.RomanAuxiliaInfantry());
        placeSyracusanUnit(hexOf(0, 7), new units.RomanHeavyInfantry());
        placeSyracusanUnit(hexOf(2, 7), new units.RomanHeavyInfantry());
        placeSyracusanUnit(hexOf(3, 7), new units.RomanHeavyInfantry());
        placeSyracusanUnit(hexOf(4, 7), new units.RomanHeavyInfantry());
        placeSyracusanUnit(hexOf(5, 7), new units.RomanAuxiliaInfantry());
        placeSyracusanUnit(hexOf(6, 7), new units.RomanAuxiliaInfantry());
        placeSyracusanUnit(hexOf(7, 7), new units.RomanMediumCavalry());

        placeSyracusanUnit(hexOf(0, 6), new units.RomanLightBowsInfantry());
        placeSyracusanUnit(hexOf(3, 6), new units.RomanLightBowsInfantry());
        placeSyracusanUnit(hexOf(7, 6), new units.RomanLightInfantry());
    }
}

const SCENARIOS = {
    akragas: new AkragasScenario(),
    melee: new MeleeScenario(),
    oneToOneMelee: new OneToOneMeleeScenario(),
    twoOnTwoMelee: new TwoOnTwoMeleeScenario(),
}

export function makeScenario(name) {
    return SCENARIOS[name] || new AkragasScenario();
}
