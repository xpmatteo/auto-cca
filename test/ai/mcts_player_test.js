import { MctsPlayer } from "ai/mcts_player.js";
import * as fs from "fs";
import makeGame from "model/game.js";
import { NullScenario, TwoOnTwoMeleeScenario } from "model/scenarios.js";
import { CarthaginianHeavyInfantry, RomanLightInfantry } from "model/units.js";
import { hexOf } from "xlib/hexlib.js";
import { fixedRandom, resetFixedRandom } from "xlib/random.js";

/**
 * @param {string} fileName
 */
function dumpTreeToFile(root, fileName) {
    const logger = fs.createWriteStream(fileName, {
        flags: 'w'
    });
    function write(str) {
        logger.write(str);
        logger.write("\n");
    }
    root.dumpTree(write);
}

const originalRandom = Math.random;
beforeEach(() => {
    resetFixedRandom();
    Math.random = fixedRandom;
});

afterEach(() => {
    Math.random = originalRandom;
});

const player = new MctsPlayer({
    expansionFactor: 2.1415,
    playoutIterations: 20,
    logfunction: () => {},
});

test('2 on 2', () => {
    const game = makeGame(new TwoOnTwoMeleeScenario());

    const root = player.search(game, 2000);

    const rootAsString = root.toString(6);
    expect(rootAsString).toMatchSnapshot();
});

test('close combat from light to heavy', () => {
    const game = makeGame(new NullScenario());
    game.placeUnit(hexOf(2, 2), new CarthaginianHeavyInfantry());
    game.placeUnit(hexOf(2, 3), new RomanLightInfantry());

    const root = player.search(game, 30);

    const rootAsString = root.toString(7);
    expect(rootAsString).toMatchSnapshot();
});
