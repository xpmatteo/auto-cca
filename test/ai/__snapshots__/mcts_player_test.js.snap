// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`2 on 2 1`] = `
"439/2000: undefined -> Roman play one card -> 1
 439/2000: PlayCard(Order Heavy Troops) -> Roman order 3 heavy units -> 1
  439/1999: End phase -> Roman movement -> 1
   439/1998: MacroCommand(Move [-1,6] to [0,5],Move [0,6] to [1,5]) -> Roman battle -> 3
    CHANCE/119: Close Combat from [0,5] to [0,4] -> Roman battle -> 2
     -38/64: Close Combat from [0,5] to [0,4] -> Carthaginian retreat -> 2
      0/0: Retreat [0,4] to [0,3] -> Roman AdvanceAfterCombat [0,5] to [0,4] -> 0
      38/64: Retreat [0,4] to [1,3] -> Roman AdvanceAfterCombat [0,5] to [0,4] -> 2
     -30/54: Close Combat from [0,5] to [0,4] -> Roman retreat -> 2
      -15/27: Retreat [0,5] to [-1,6] -> Roman AdvanceAfterCombat [0,4] to [0,5] -> 2
      -15/27: Retreat [0,5] to [0,6] -> Roman AdvanceAfterCombat [0,4] to [0,5] -> 2
    CHANCE/593: Close Combat from [1,5] to [1,4] -> Roman battle -> 6
     -43/72: Close Combat from [1,5] to [1,4] -> Carthaginian retreat -> 2
      43/72: Retreat [1,4] to [2,3] -> Roman AdvanceAfterCombat [1,5] to [1,4] -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman AdvanceAfterCombat [1,5] to [1,4] -> 0
     17/79: Close Combat from [1,5] to [1,4] -> Roman retreat -> 2
      16/54: Retreat [1,5] to [0,6] -> Roman AdvanceAfterCombat [1,4] to [1,5] -> 2
      1/25: Retreat [1,5] to [1,6] -> Roman AdvanceAfterCombat [1,4] to [1,5] -> 2
     -62/240: Close Combat from [1,5] to [1,4] -> Carthaginian retreat -> 2
      62/240: Retreat [1,4] to [2,3] -> Roman AdvanceAfterCombat [1,5] to [1,4] -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman AdvanceAfterCombat [1,5] to [1,4] -> 0
     -60/62: Close Combat from [1,5] to [1,4] -> Roman retreat -> 2
      -31/33: Retreat [1,5] to [0,6] -> Roman AdvanceAfterCombat [1,4] to [1,5] -> 2
      -29/29: Retreat [1,5] to [1,6] -> Roman AdvanceAfterCombat [1,4] to [1,5] -> 2
     47/73: Close Combat from [1,5] to [1,4] -> Roman retreat -> 2
      16/29: Retreat [1,5] to [0,6] -> Roman AdvanceAfterCombat [1,4] to [1,5] -> 2
      31/44: Retreat [1,5] to [1,6] -> Roman AdvanceAfterCombat [1,4] to [1,5] -> 2
     -61/66: Close Combat from [1,5] to [1,4] -> Carthaginian retreat -> 2
      61/66: Retreat [1,4] to [2,3] -> Roman AdvanceAfterCombat [1,5] to [1,4] -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman AdvanceAfterCombat [1,5] to [1,4] -> 0
    CHANCE/1287: Close Combat from [0,5] to [1,4] -> Roman battle -> 5
     144/465: Close Combat from [0,5] to [1,4] -> Roman retreat -> 2
      112/317: Retreat [0,5] to [-1,6] -> Roman AdvanceAfterCombat [1,4] to [0,5] -> 2
      32/148: Retreat [0,5] to [0,6] -> Roman AdvanceAfterCombat [1,4] to [0,5] -> 2
     -73/401: Close Combat from [0,5] to [1,4] -> Carthaginian retreat -> 2
      73/401: Retreat [1,4] to [2,3] -> Roman AdvanceAfterCombat [0,5] to [1,4] -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman AdvanceAfterCombat [0,5] to [1,4] -> 0
     -52/80: Close Combat from [0,5] to [1,4] -> Carthaginian retreat -> 2
      52/80: Retreat [1,4] to [2,3] -> Roman AdvanceAfterCombat [0,5] to [1,4] -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman AdvanceAfterCombat [0,5] to [1,4] -> 0
     -129/189: Close Combat from [0,5] to [1,4] -> Carthaginian retreat -> 2
      129/189: Retreat [1,4] to [2,3] -> Roman AdvanceAfterCombat [0,5] to [1,4] -> 2
      0/0: Retreat [1,4] to [1,3] -> Roman AdvanceAfterCombat [0,5] to [1,4] -> 0
     -137/151: Close Combat from [0,5] to [1,4] -> Roman retreat -> 2
      -67/73: Retreat [0,5] to [-1,6] -> Roman AdvanceAfterCombat [1,4] to [0,5] -> 2
      -70/78: Retreat [0,5] to [0,6] -> Roman AdvanceAfterCombat [1,4] to [0,5] -> 2
"
`;

exports[`close combat from light to heavy 1`] = `
"-26/30: undefined -> Roman play one card -> 1
 -26/30: PlayCard(Order Light Troops) -> Roman order 1 light units -> 1
  -25/29: End phase -> Roman movement -> 1
   -25/28: MacroCommand(Move [2,3] to [3,2]) -> Roman battle -> 1
    CHANCE/27: Close Combat from [3,2] to [2,2] -> Roman battle -> 5
     5/5: Close Combat from [3,2] to [2,2] -> Carthaginian retreat -> 2
      -5/5: Retreat [2,2] to [3,1] -> Roman AdvanceAfterCombat [3,2] to [2,2] -> 2
       -2/2: Retreat [3,2] to [2,2] -> Roman battle -> 1
       -2/2: Retreat [3,2] to [3,2] -> Roman battle -> 1
      0/0: Retreat [2,2] to [2,1] -> Roman AdvanceAfterCombat [3,2] to [2,2] -> 0
     -7/7: Close Combat from [3,2] to [2,2] -> Roman battle -> 1
      7/7: End phase -> Carthaginian play one card -> 1
       6/6: PlayCard(Order Heavy Troops) -> Carthaginian order 1 heavy units -> 1
     -5/5: Close Combat from [3,2] to [2,2] -> Roman retreat -> 3
      -2/2: Retreat [3,2] to [3,4] -> Roman AdvanceAfterCombat [2,2] to [3,2] -> 2
       -2/2: Retreat [2,2] to [3,2] -> Roman battle -> 1
       0/0: Retreat [2,2] to [2,2] -> Roman battle -> 0
      -1/1: Retreat [3,2] to [1,4] -> Roman AdvanceAfterCombat [2,2] to [3,2] -> 2
       -1/1: Retreat [2,2] to [3,2] -> Roman battle -> 0
       0/0: Retreat [2,2] to [2,2] -> Roman battle -> 0
      -2/2: Retreat [3,2] to [2,4] -> Roman AdvanceAfterCombat [2,2] to [3,2] -> 2
       -1/1: Retreat [2,2] to [3,2] -> Roman battle -> 0
       0/0: Retreat [2,2] to [2,2] -> Roman battle -> 0
     0/3: Close Combat from [3,2] to [2,2] -> Carthaginian retreat -> 2
      0/0: Retreat [2,2] to [3,1] -> Roman AdvanceAfterCombat [3,2] to [2,2] -> 0
      0/3: Retreat [2,2] to [2,1] -> Roman AdvanceAfterCombat [3,2] to [2,2] -> 2
       0/1: Retreat [3,2] to [2,2] -> Roman battle -> 0
       1/1: Retreat [3,2] to [3,2] -> Roman battle -> 1
     -6/6: Close Combat from [3,2] to [2,2] -> Roman retreat -> 3
      -2/2: Retreat [3,2] to [1,4] -> Roman AdvanceAfterCombat [2,2] to [3,2] -> 2
       0/0: Retreat [2,2] to [3,2] -> Roman battle -> 0
       -1/1: Retreat [2,2] to [2,2] -> Roman battle -> 0
      -2/2: Retreat [3,2] to [2,4] -> Roman AdvanceAfterCombat [2,2] to [3,2] -> 2
       -2/2: Retreat [2,2] to [2,2] -> Roman battle -> 1
       0/0: Retreat [2,2] to [3,2] -> Roman battle -> 0
      -2/2: Retreat [3,2] to [3,4] -> Roman AdvanceAfterCombat [2,2] to [3,2] -> 2
       -2/2: Retreat [2,2] to [2,2] -> Roman battle -> 1
       0/0: Retreat [2,2] to [3,2] -> Roman battle -> 0
"
`;
